pkg/beaconfqdn/mongodb.go

```diff
@@ -151,8 +178,189 @@ func (r *repo) Upsert(hostnameMap map[string]*hostname.Input, minTimestamp, maxT
 		bar.IncrBy(1)
 
 	}
+
 	p.Wait()
 
 	// start the closing cascade (this will also close the other channels)
 	dissectorWorker.close()
 }
+
+// affectedHostnameIPs gathers all of the hostnames associated with the external IPs which generated
+// traffic in this run. Each hostname entry is returned along with its list of associated resolved IPs.
+func (r *repo) affectedHostnameIPs(hostMap map[string]*host.Input) ([]hostnameIPs, error) {
+
+	// We can only submit around 200,000 hosts in a single query to MongoDB due to the 16MB document limit.
+	// 1 Megabyte / Size of Unique IP BSON query ~= 200,000
+	// 16 * 1024 * 1024 / (64 for IPv4, 83 for IPv6) = (262144 for IPv4, 202135 for IPv6)
+
+	// provide a fast path for datasets with less than 200,000 hosts
+	if len(hostMap) <= 200000 {
+		return r.affectedHostnameIPsSimple(hostMap)
+	}
+	return r.affectedHostnameIPsChunked(hostMap)
+}
+
+// affectedHostnameIPsSimple implements affectedHostnameIPs but should only be called with hostMaps with
+// roughly less than 200,000 external hosts.
+func (r *repo) affectedHostnameIPsSimple(hostMap map[string]*host.Input) ([]hostnameIPs, error) {
+	// preallocate externalHosts slice assuming at least half of the observed hosts are external
+	var externalHosts []data.UniqueIP = make([]data.UniqueIP, 0, len(hostMap)/2)
+	var affectedHostnamesBuffer []hostnameIPs
+
+	ssn := r.database.Session.Copy()
+	defer ssn.Close()
+
+	for _, host := range hostMap {
+		if host.IsLocal {
+			continue
+		}
+		externalHosts = append(externalHosts, host.Host)
+	}
+	reverseDNSagg := reverseDNSQueryWithIPs(externalHosts)
+	externalHosts = nil // nolint ,we are freeing this potentially large array so the GC can claim it during MongoDB IO
+
+	err := ssn.DB(r.database.GetSelectedDB()).C(r.config.T.DNS.HostnamesTable).
+		Pipe(reverseDNSagg).AllowDiskUse().All(&affectedHostnamesBuffer)
+	return affectedHostnamesBuffer, err
+}
+
+// affectedHostnameIPsChunked implements affectedHostnameIPs. It is slower than affectedHostnameIPs_simple
+// and uses more RAM. However, unlike affectedHostnameIPs_simple, affectedHostnameIPs_chunked handles
+// hostMaps of all sizes.
+func (r *repo) affectedHostnameIPsChunked(hostMap map[string]*host.Input) ([]hostnameIPs, error) {
+	// preallocate externalHosts slice assuming at least half of the observed hosts are external
+	// util.Min ensures that we don't preallocate more than the maximum allowed number of hosts in a 16MB BSON document
+	var externalHosts []data.UniqueIP = make([]data.UniqueIP, 0, util.Min(200000, len(hostMap)/2))
+	var affectedHostnamesBuffer []hostnameIPs
+
+	ssn := r.database.Session.Copy()
+	defer ssn.Close()
+
+	// we will need to remove duplicate results from each query of 200,000 hosts, slowing down the process
+	// and consuming more RAM
+
+	// affectedHostnameIPMap maps hostnames to their respective ResolvedIPs
+	var affectedHostnameIPMap map[string][]data.UniqueIP = make(map[string][]data.UniqueIP, len(hostMap)/10)
```

Comment:
Maybe quick comment here explaining the /10 factor.

Reply:
This assumes that there would be roughly 10 IPs per observed hostname in the dataset. I'm not sure that's a good assumption to be honest. I didn't want to pre-allocate with `len(hostMap)` since that would obviously be too big. But, I also didn't want to start from 0. Thoughts?

Reply:
I think 10 is reasonable. Maybe just add a comment mentioning that?

Reply:
Oh hang on though, this wouldn't be 10 IPs per host. This would be len(hostMap)/10 IPs per host.

Reply:
len(hostMap)/10 per hostname* or rather #IPs/10 : #FQDNs. 

So, this assumes that for every 10 IPs in the parsed in set of logs, there would be 1 FQDN.

Reply:
Got it, was reading the make statement wrong!